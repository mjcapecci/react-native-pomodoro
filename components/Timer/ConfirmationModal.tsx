import * as React from 'react'
import { Modal, Portal, Text, Button } from 'react-native-paper'
import { TimerContext } from './TimerContextProviderV2'
import styles from '../General/General.style'
import ModalDivider from '../General/ModalDivider'

interface ConfirmationModalProps {
  skipType: 'skip' | 'stop'
  showModal: boolean
  setShowModal: (showModal: boolean) => void
  secondsLeft: number
}

const ConfirmationModal = ({
  skipType,
  showModal,
  setShowModal,
  secondsLeft,
}: ConfirmationModalProps): JSX.Element => {
  const time = React.useContext(TimerContext)

  const hideModal = (): void => setShowModal(false)
  const containerStyle = { backgroundColor: 'white', padding: 20 }

  const confirmSkip = (): void => {
    if (skipType === 'stop') {
      time.stopRound()
    } else {
      time.stopRound(true)
    }
    hideModal()
  }

  const disabled = secondsLeft < 3

  return (
    <Portal>
      <Modal visible={showModal} onDismiss={hideModal} contentContainerStyle={containerStyle}>
        <Text style={styles.centeredModalText}>
          Are you sure you would like to skip this round?
        </Text>
        <ModalDivider />
        <Button onPress={() => hideModal()}>Cancel</Button>

        <ModalDivider />
        <Button
          style={styles.bottomButton}
          color={disabled ? 'grey' : '#ed0d0dc4'}
          onPress={() => !disabled && confirmSkip()}
        >
          Skip {disabled && '(not available)'}
        </Button>
      </Modal>
    </Portal>
  )
}

export default ConfirmationModal
