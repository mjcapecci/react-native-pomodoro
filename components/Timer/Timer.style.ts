import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#000',
  },
  timeHeaderContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    height: 100,
  },
  timeHeader: {
    fontSize: 84,
    color: '#fff',
    marginBottom: 5,
    fontVariant: ['tabular-nums'],
    borderLeftWidth: 4,
    borderRightWidth: 4,
  },
  starContainer: {
    display: 'flex',
    flexDirection: 'row',
  },
  star: {
    margin: 3,
  },
  actionButton: {
    marginTop: 60,
    background: 'none',
  },
  skipButton: {
    marginTop: 30,
    background: 'none',
  },
  hideSkip: {
    display: 'none',
  },
  loadingCaption: {
    color: '#fff',
    marginBottom: 25,
    fontSize: 24,
  },
})
