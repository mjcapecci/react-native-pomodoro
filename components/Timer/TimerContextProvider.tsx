import React, { createContext, useCallback, useEffect, useState } from 'react'
import { getNextRound, getRoundType, getSecondsReset } from './helpers/timerHelpers'
import { scheduleNotification, cancelAllNotifications } from '../Notifications/notificationManager'
import AsyncStorage from '@react-native-async-storage/async-storage'
import { AppState, AppStateStatus } from 'react-native'
import getTimeRemaining, { getRoundData } from './helpers/getTimeRemaining'
import { RoundData, RoundType } from '../../types'
import addRecord from '../../data_layer/addRecord'
import shouldAddRecord from './helpers/shouldAddRecord'
import { useMutation, useQueryClient } from '@tanstack/react-query'

export interface TimerContextProps {
  enabled: boolean
  startButtonEnabled: boolean
  timerActive: boolean
  roundNumber: number
  roundType: string
  secondsLeft: number
  appStateVisible: boolean
  advanceRound: () => void
  startRound: () => void
  stopRound: () => void
  showConfirmationModal: boolean
  setShowConfirmationModal: (show: boolean) => void
}

const TimerContext = createContext<TimerContextProps>({
  enabled: true,
  startButtonEnabled: true,
  timerActive: false,
  roundNumber: 0,
  roundType: 'work',
  secondsLeft: 1500,
  appStateVisible: true,
  advanceRound: () => null,
  startRound: () => null,
  stopRound: () => null,
  showConfirmationModal: false,
  setShowConfirmationModal: () => null,
})

interface TimerContextProviderProps {
  children: JSX.Element
}

function TimerContextProvider({ children }: TimerContextProviderProps): JSX.Element {
  const [enabled, setEnabled] = useState(true)
  const [startButtonEnabled, setStartButtonEnabled] = useState(true)
  const [timerActive, setTimerActive] = useState(false)
  const [roundNumber, setRoundNumber] = useState(0)
  const [roundType, setRoundType] = useState<RoundType>(RoundType.Work)
  const [secondsLeft, setSecondsLeft] = useState(1500)
  const [appStateVisible, setAppStateVisible] = useState(true)
  const [showConfirmationModal, setShowConfirmationModal] = useState(false)

  // React Query
  const queryClient = useQueryClient()

  const mutation = useMutation(addRecord, {
    onSuccess: () => {
      void queryClient.invalidateQueries(['records'])
    },
  })

  // ----- TIMER CONTROL METHODS -----
  const advanceRound = useCallback((): void => {
    const nextRound = getNextRound(roundNumber)
    const nextRoundType = getRoundType(nextRound)
    setRoundNumber(nextRound)
    setRoundType(nextRoundType)
    setSecondsLeft(getSecondsReset(nextRoundType))
  }, [roundNumber])

  const startTimer = async (): Promise<void> => {
    const roundData: RoundData = {
      date: new Date().getTime(),
      roundNumber,
      roundType,
    }

    await AsyncStorage.setItem('roundData', JSON.stringify(roundData))
    await scheduleNotification(secondsLeft)
  }

  const startRound = async (): Promise<void> => {
    if (enabled && startButtonEnabled) {
      setTimerActive(true)
      setSecondsLeft(secondsLeft - 1)
      await startTimer()
    }
  }

  const stopRound = async (): Promise<void> => {
    setTimerActive(false)
    setEnabled(false)
    await stopTimer()
  }

  const stopTimer = async (): Promise<void> => {
    const roundData: RoundData | undefined = await getRoundData()

    if (timerActive) {
      advanceRound()
    }

    if (await shouldAddRecord(roundData?.date ?? 0)) {
      await mutation.mutateAsync({
        date: roundData?.date ?? 0,
        type: roundData?.roundType ?? RoundType.Work,
        completed: secondsLeft < 0 ? 1 : 0,
      })
    }

    await AsyncStorage.removeItem('roundData')
    await cancelAllNotifications()
    await setTimeout(() => {
      setSecondsLeft(getSecondsReset(getRoundType(getNextRound(roundNumber - 1))))
      setEnabled(true)
    }, 1000)
  }

  // This effect adds a slight delay to the start button to prevent a known race condition
  useEffect(() => {
    if (enabled) {
      setTimeout(() => {
        setStartButtonEnabled(true)
      }, 1000)
    } else {
      setStartButtonEnabled(false)
    }
  }, [enabled])

  // This effect updates the display time every second
  useEffect(() => {
    const fetchTime = async (): Promise<number> => {
      return await getTimeRemaining()
    }

    if (timerActive && secondsLeft > -1) {
      setTimeout(async () => {
        setSecondsLeft((await fetchTime()) ?? secondsLeft)
      }, 1000)
    } else {
      void (async () => {
        await stopTimer()
      })()
      setTimerActive(false)
      setEnabled(false)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [secondsLeft])

  const handleAppStateChange = useCallback(
    async (nextAppState: AppStateStatus): Promise<void> => {
      if (nextAppState === 'active') {
        if (!appStateVisible) {
          setTimeout(async () => {
            setSecondsLeft((await getTimeRemaining()) ?? secondsLeft)
            setAppStateVisible(true)
          }, 1000)
        }
      }

      if (nextAppState === 'inactive' || nextAppState === 'background') {
        return setAppStateVisible(false)
      }
    },
    [appStateVisible, secondsLeft],
  )

  // Detecting changes in AppState
  useEffect(() => {
    const subscription = AppState.addEventListener(
      'change',
      async (nextAppState) => await handleAppStateChange(nextAppState),
    )

    return () => {
      subscription.remove()
    }
  }, [handleAppStateChange])

  return (
    <TimerContext.Provider
      value={{
        enabled,
        startButtonEnabled,
        timerActive,
        roundNumber,
        roundType,
        secondsLeft,
        appStateVisible,
        advanceRound,
        startRound,
        stopRound,
        showConfirmationModal,
        setShowConfirmationModal,
      }}
    >
      {children}
    </TimerContext.Provider>
  )
}

export { TimerContext, TimerContextProvider }
