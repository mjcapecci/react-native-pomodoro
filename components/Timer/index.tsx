import React, { useState, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { StyleSheet, View } from 'react-native';
import { Heading, Box, Button } from 'native-base';
import Ionicons from '@expo/vector-icons/Ionicons';
import {
  scheduleNotification,
  cancelAllNotifications,
} from '../Notifications/notificationManager';
import {
  fmtMSS,
  getIconColor,
  getNextRound,
  getNextRoundSecondsDisplay,
  getRoundType,
  getSecondsReset,
} from '../../utils/TimerUtils';

const Timer = () => {
  const [enabled, setEnabled] = useState(true);
  const [startButtonEnabled, setStartButtonEnabled] = useState(true);
  const [timerActive, setTimerActive] = useState(false);
  const [roundNumber, setRoundNumber] = useState(-1);
  const [roundType, setRoundType] = useState('work');
  const [secondsLeft, setSecondsLeft] = useState(1500);

  const debugMode = false;

  useEffect(() => {
    if (!timerActive) {
      advanceRound();
    }
  }, [timerActive]);

  useEffect(() => {
    if (enabled) {
      setTimeout(() => {
        setStartButtonEnabled(true);
      }, 1000);
    } else {
      setStartButtonEnabled(false);
    }
  }, [enabled]);

  useEffect(() => {
    if (timerActive && secondsLeft > -1) {
      setTimeout(() => {
        setSecondsLeft(secondsLeft - 1);
      }, 1000);
    } else {
      stopTimer();
      setTimerActive(false);
      setEnabled(false);
    }
  }, [secondsLeft]);

  const advanceRound = () => {
    const nextRound = getNextRound(roundNumber);
    const nextRoundType = getRoundType(nextRound);
    setRoundNumber(nextRound);
    setRoundType(nextRoundType);
    setSecondsLeft(getSecondsReset(nextRoundType));
  };

  const startTimer = async () => {
    const roundData = {
      date: new Date().getTime(),
      roundNumber: roundNumber,
      roundType: roundType,
    };

    await AsyncStorage.setItem('roundData', JSON.stringify(roundData));
    await scheduleNotification(secondsLeft);
  };

  const stopTimer = async () => {
    await AsyncStorage.removeItem('roundData');
    await cancelAllNotifications();
    await setTimeout(() => {
      setSecondsLeft(
        getSecondsReset(getRoundType(getNextRound(roundNumber - 1)))
      );
      setEnabled(true);
    }, 1000);
  };

  const printAsyncStorage = async () => {
    const startTime = await AsyncStorage.getItem('roundData');
    console.log(startTime);
  };

  return enabled ? (
    <View style={styles.container}>
      <Box style={styles.timeHeaderContainer}>
        <Heading style={styles.timeHeader}>
          {timerActive
            ? secondsLeft > 0
              ? fmtMSS(secondsLeft)
              : '0:00'
            : getNextRoundSecondsDisplay(roundNumber)}
        </Heading>
      </Box>
      <Box style={styles.starContainer}>
        <Ionicons
          name='hammer'
          color={getIconColor(0, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
        <Ionicons
          name='walk'
          color={getIconColor(1, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
        <Ionicons
          name='hammer'
          color={getIconColor(2, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
        <Ionicons
          name='walk'
          color={getIconColor(3, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
        <Ionicons
          name='hammer'
          color={getIconColor(4, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
        <Ionicons
          name='pizza'
          color={getIconColor(5, roundNumber, timerActive)}
          size={24}
          style={styles.star}
        />
      </Box>
      <Box style={styles.actionButton}>
        {!timerActive ? (
          <Ionicons
            name='play-circle-outline'
            color={startButtonEnabled ? 'white' : 'grey'}
            size={104}
            onPress={() => {
              if (enabled && startButtonEnabled) {
                setTimerActive(true);
                setSecondsLeft(secondsLeft - 1);
                startTimer();
              }
            }}
          ></Ionicons>
        ) : (
          <Ionicons
            name='stop-circle-outline'
            color='white'
            size={104}
            onPress={() => {
              setTimerActive(false);
              setEnabled(false);
              stopTimer();
            }}
          ></Ionicons>
        )}
      </Box>
      <Button style={!timerActive ? styles.skipButton : styles.hideSkip}>
        {' '}
        <Ionicons
          name='play-skip-forward-circle'
          color='white'
          size={50}
          onPress={() => advanceRound()}
        ></Ionicons>
      </Button>
      {debugMode && (
        <Button>
          {' '}
          <Ionicons
            name='play-skip-forward-circle'
            color='white'
            size={50}
            onPress={() => printAsyncStorage()}
          ></Ionicons>
        </Button>
      )}
    </View>
  ) : (
    <View style={styles.container}>
      <Box style={styles.timeHeaderContainer}>
        <Heading>Loading...</Heading>
      </Box>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#000',
  },
  timeHeaderContainer: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
  },
  timeHeader: {
    fontSize: 84,
    color: '#fff',
    marginBottom: 5,
    fontVariant: ['tabular-nums'],
  },
  starContainer: {
    display: 'flex',
    flexDirection: 'row',
  },
  star: {
    margin: 3,
  },
  actionButton: {
    marginTop: 60,
  },
  skipButton: {
    marginTop: 30,
    backgroundColor: '#000',
  },
  hideSkip: {
    display: 'none',
  },
});

export default Timer;
