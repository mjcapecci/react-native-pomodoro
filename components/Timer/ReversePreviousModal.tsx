import * as React from 'react'
import { Modal, Portal, Text, Button } from 'react-native-paper'
import { TimerContext } from './TimerContextProvider'

interface ReversePreviousModalProps {
  showModal: boolean
  setShowModal: (showModal: boolean) => void
}

const ReversePreviousModal = ({
  showModal,
  setShowModal,
}: ReversePreviousModalProps): JSX.Element => {
  const time = React.useContext(TimerContext)

  const hideModal = (): void => setShowModal(false)
  const containerStyle = { backgroundColor: 'white', padding: 20 }

  const confirmReverse = async (): Promise<void> => {
    if (time?.lastUserRecord?.id != null) {
      await time.reversePreviousRecord(time.lastUserRecord.id)
    }
    hideModal()
  }

  return (
    <Portal>
      <Modal visible={showModal} onDismiss={hideModal} contentContainerStyle={containerStyle}>
        <Text>
          Are you sure you would like to set the last completed round to a status of skipped?
        </Text>
        <Button onPress={async () => await confirmReverse()}>Skip</Button>
        <Button onPress={() => hideModal()}>Cancel</Button>
      </Modal>
    </Portal>
  )
}

export default ReversePreviousModal
