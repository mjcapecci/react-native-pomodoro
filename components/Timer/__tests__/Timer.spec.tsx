import React from 'react';
import Timer from '..';
import { render, cleanup, act } from '@testing-library/react-native';
import {
  TimerContext,
  TimerContextProps,
  TimerContextProvider,
} from '../TimerContextProvider';
import { activeRound } from '../../../__mocks__/fixtures/Timer.fixture';

afterEach(() => cleanup());

function mountForSnap(value: TimerContextProps) {
  return render(
    <TimerContext.Provider value={value}>
      <Timer />
    </TimerContext.Provider>
  );
}

function mountForAction() {
  return render(
    <TimerContextProvider>
      <Timer />
    </TimerContextProvider>
  );
}

describe('Timer Component', () => {
  it('renders properly on mount', () => {
    const tree = render(<Timer />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders properly when active', async () => {
    const tree = mountForSnap(activeRound).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can switch from inactive to active', async () => {
    jest.useFakeTimers();
    const tree = mountForAction().toJSON();

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(tree).toMatchSnapshot();
  });
});
