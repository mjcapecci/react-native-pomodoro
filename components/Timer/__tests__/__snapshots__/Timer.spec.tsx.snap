// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer Component can switch from inactive to active 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": 100,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Object {
            "color": "#fff",
            "fontSize": 24,
            "marginBottom": 25,
          },
        ]
      }
    >
      Loading work round...
    </Text>
    <ActivityIndicator
      color="#fff"
      size="large"
    />
  </View>
</View>
`;

exports[`Timer Component renders properly on mount 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="play-button"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": 100,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Object {
            "borderColor": "blue",
            "borderLeftWidth": 4,
            "borderRightWidth": 4,
            "color": "#fff",
            "fontSize": 84,
            "fontVariant": Array [
              "tabular-nums",
            ],
            "marginBottom": 5,
          },
        ]
      }
      testID="active-time-header"
    >
      invalid_round_number
    </Text>
  </View>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <Text />
    <Text />
    <Text />
    <Text />
    <Text />
    <Text />
  </View>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "transparent",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 0,
        "marginTop": 60,
        "minWidth": 64,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
      testID="main-action-button"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          <Text />
        </Text>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "#000",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 0,
        "marginTop": 30,
        "minWidth": 64,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
           
          <Text />
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Timer Component renders properly when active 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="play-button"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": 100,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Object {
            "borderColor": "blue",
            "borderLeftWidth": 4,
            "borderRightWidth": 4,
            "color": "#fff",
            "fontSize": 84,
            "fontVariant": Array [
              "tabular-nums",
            ],
            "marginBottom": 5,
          },
        ]
      }
      testID="active-time-header"
    >
      25:00
    </Text>
  </View>
  <View
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "cyan",
            "fontSize": 24,
          },
          Object {
            "margin": 3,
          },
          Object {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "grey",
            "fontSize": 24,
          },
          Object {
            "margin": 3,
          },
          Object {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "grey",
            "fontSize": 24,
          },
          Object {
            "margin": 3,
          },
          Object {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "grey",
            "fontSize": 24,
          },
          Object {
            "margin": 3,
          },
          Object {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "grey",
            "fontSize": 24,
          },
          Object {
            "margin": 3,
          },
          Object {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "grey",
            "fontSize": 24,
          },
          Object {
            "margin": 3,
          },
          Object {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "transparent",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 0,
        "marginTop": 60,
        "minWidth": 64,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
      testID="main-action-button"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "white",
                  "fontSize": 104,
                },
                undefined,
                Object {
                  "fontFamily": "ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="stop-button-icon"
          >
            
          </Text>
        </Text>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "transparent",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "display": "none",
        "elevation": 0,
        "minWidth": 64,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#6200ee",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
           
          <Text
            allowFontScaling={false}
            onPress={[Function]}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "white",
                  "fontSize": 50,
                },
                undefined,
                Object {
                  "fontFamily": "ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </Text>
      </View>
    </View>
  </View>
</View>
`;
